ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManager.java

```diff
@@ -543,6 +582,70 @@ public void onInstanceConfigChange(List<InstanceConfig> configs, NotificationCon
       }
     }
 
+    /**
+     * Triggered whenever the IdealState in current data center has changed (for now, it is usually updated by Helix Bootstrap tool).
+     * @param idealState a list of {@link IdealState} that specifies ideal location of replicas.
+     * @param changeContext the {@link NotificationContext} associated.
+     * @throws InterruptedException
+     */
+    @Override
+    public void onIdealStateChange(List<IdealState> idealState, NotificationContext changeContext)
+        throws InterruptedException {
+      if (!idealStateInitialized.get()) {
+        logger.info("Received initial notification for IdealState change from {}", dcName);
+        idealStateInitialized.set(true);
+      } else {
+        logger.info("IdealState change triggered from {}", dcName);
+      }
+      synchronized (notificationLock) {
```

Comment:
so looks like notificationlock is being used to protect multiple unrelated data. I see 4 usages

1. to protect instanceNameToAmbryDataNode, dcToNode, allInstances
2. to protect partitionToResourceNameByDc
3. to protect dcToRoutingTableSnapshotRef
4. to protect instanceNameToAmbryDatanode

We do not need to protect all of them by the same lock. For last three cases we can use the synchronized block with the object itself e.g, synchronized(instanceNameToAmbryDatanode) {}

Reply:
Right, I think we don't have to add additional lock for `partitionToResourceNameByDc` and `dcToRoutingTableSnapshotRef`. The cluster changes are grouped by datacenter and these two concurrent hashmaps should be able to support concurrent changes in different data centers.

Reply:
Sorry, you are right, `partitionToResourceNameByDc` should be synchronized